src/Features/Core/Portable/Extensions/ExtensionFolder.cs

```diff
@@ -0,0 +1,172 @@
+                        // The file is not a valid .NET assembly, skip it.
+                        continue;
+                    }
+
+                    analyzerAssemblyLoader.AddDependencyLocation(dll);
+                }
+
+                return analyzerAssemblyLoader;
+            });
+        }
+
+        private async Task<AssemblyMessageHandlers> CreateAssemblyHandlersAsync(
+            string assemblyFilePath, CancellationToken cancellationToken)
+        {
+            // If creating the underlying assembly loader failed, then we will throw that exception outwards for the
+            // client to hear about.
+            var analyzerAssemblyLoader = await _lazyAssemblyLoader.GetValueAsync(cancellationToken).ConfigureAwait(false);
+
+            var assembly = analyzerAssemblyLoader.LoadFromPath(assemblyFilePath);
+            var factory = _extensionMessageHandlerService.CustomMessageHandlerFactory;
+
+            var documentMessageHandlers = factory
+                .CreateDocumentMessageHandlers(assembly, extensionIdentifier: assemblyFilePath, cancellationToken)
+                .ToImmutableDictionary(h => h.Name, h => (IExtensionMessageHandlerWrapper)h);
+            var workspaceMessageHandlers = factory
+                .CreateWorkspaceMessageHandlers(assembly, extensionIdentifier: assemblyFilePath, cancellationToken)
+                .ToImmutableDictionary(h => h.Name, h => (IExtensionMessageHandlerWrapper)h);
+
+            return new(documentMessageHandlers, workspaceMessageHandlers);
+        }
+
+        public void RegisterAssembly(string assemblyFilePath)
+        {
+            lock (_assemblyFilePathToHandlers_useOnlyUnderLock)
+            {
+                if (_assemblyFilePathToHandlers_useOnlyUnderLock.ContainsKey(assemblyFilePath))
+                    throw new InvalidOperationException($"Extension '{assemblyFilePath}' is already registered.");
+
+                _assemblyFilePathToHandlers_useOnlyUnderLock.Add(
+                    assemblyFilePath,
+                    AsyncLazy.Create(
+                        cancellationToken => this.CreateAssemblyHandlersAsync(assemblyFilePath, cancellationToken)));
+            }
+        }
+
+        /// <summary>
+        /// Unregisters this assembly path from this extension folder.  If this was the last registered path, then this
+        /// will return true so that this folder can be unloaded.
+        /// </summary>
+        public bool UnregisterAssembly(string assemblyFilePath)
+        {
+            lock (_assemblyFilePathToHandlers_useOnlyUnderLock)
+            {
+                if (!_assemblyFilePathToHandlers_useOnlyUnderLock.ContainsKey(assemblyFilePath))
+                    throw new InvalidOperationException($"Extension '{assemblyFilePath}' was not registered.");
+
+                _assemblyFilePathToHandlers_useOnlyUnderLock.Remove(assemblyFilePath);
+                return _assemblyFilePathToHandlers_useOnlyUnderLock.Count == 0;
```

Comment:
are we potentially leaking ALCs here?  Do we need to dispose of the aalyzer loader if there are no more extensions for the folder?

Reply:
while the GC should handle this for us, i added some explicit unloading to make lifetimes clearer.